nanofi/src/core/flowfiles.c

```diff
@@ -0,0 +1,169 @@
+    snprintf(offset_str, sizeof(offset_str), "%llu", curr_offset);
+    update_attribute(ffr, "current offset", offset_str, strlen(offset_str));
+    char content_location[strlen(ffr->contentLocation) + 1];
+    snprintf(content_location, sizeof(content_location), "%s", ffr->contentLocation);
+    update_attribute(ffr, "content location", content_location, strlen(content_location));
+    update_attribute(ffr, "tailfile path", (char*)file_path, strlen(file_path));
+}
+
+void transmit_flow_files(nifi_instance * instance, flow_file_list * ff_list, int complete) {
+    if (!instance || !ff_list) {
+        return;
+    }
+    flow_file_list * el = NULL;
+    LL_FOREACH(ff_list, el) {
+        if (el) {
+            if (!complete) {
+                transmit_flowfile(el->ff_record, instance);
+            }
+            if (complete && el->complete) {
+                transmit_flowfile(el->ff_record, instance);
+            }
+        }
+    }
+}
+
+void read_payload_and_transmit(struct flow_file_list * ffl, struct CRawSiteToSiteClient * client) {
+    if (!ffl || !client) {
+        return;
+    }
+
+    char * file = ffl->ff_record->contentLocation;
+    FILE * fp = fopen(file, "rb");
+
+    struct stat statfs;
+    stat(file, &statfs);
+    size_t file_size = statfs.st_size;
+
+    attribute attr;
+    attr.key = "current offset";
+    if (get_attribute(ffl->ff_record, &attr) < 0) {
+        printf("Error looking up flow file attribute %s\n", attr.key);
+        return;
+    }
+
+    errno = 0;
+    uint64_t offset = strtoull((const char *)attr.value, NULL, 10);
+    if (errno != 0) {
+        printf("Error converting flow file offset value\n");
+        return;
+    }
+    uint64_t begin_offset =  offset - file_size;
+    if (fp) {
+        char * buff = (char *)malloc(sizeof(char) * 4097);
+        size_t count = 0;
+        while ((count = fread(buff, 1, 4096, fp)) > 0) {
+            buff[count] = '\0';
+            begin_offset += count;
+            char offset_str[21];
+            snprintf(offset_str, sizeof(offset_str), "%llu", begin_offset);
+            update_attribute(ffl->ff_record, "current offset", offset_str, strlen(offset_str));
+
+            attribute_set as;
+            uint64_t num_attrs = get_attribute_quantity(ffl->ff_record);
+            as.size = num_attrs;
+            as.attributes = (attribute *)malloc(num_attrs * sizeof(attribute));
+            get_all_attributes(ffl->ff_record, &as);
+
+            if (transmitPayload(client, buff, &as) == 0) {
+                printf("payload of %zu bytes from %s sent successfully\n", count, ffl->ff_record->contentLocation);
+            }
+            else {
+                printf("Failed to send payload, flow file %s\n", ffl->ff_record->contentLocation);
+            }
+        }
+        free(buff);
+    }
+    fclose(fp);
+}
+
+void transmit_payload(struct CRawSiteToSiteClient * client, struct flow_file_list * ff_list, int complete) {
+    if (!client || !ff_list) {
```

Comment:
The null check macro could be used here, too.

Reply:
I understand that NULL macro can be used but this is more readable and straight forward.